#11081 Add C/C++ section to Noteworthy differences documentation

ScottPJones opened the pull request:
Updated doc/manual/noteworthy-differences.rst to have a section of a number of the differences between C/C++ and Jullia that I found confusion initally, coming from a C/C++ background.

tkelman commented:
You're mixing up code changes from other PR's with documentation changes. I'm guessing you only meant for the single documentation commit to be included here?

ScottPJones commented:
@tkelman yes, of course, I must have done something wrong creating the branch, pushing it back to github, or something... I have no idea how it happened, or how to fix it... this should just have the one commit, 5f5f3c7 with just the one file changed (no edits, just appending new section), to fix #11081 (which should be re-opened until this is accepted!)

timholy commented:
One way to do this is to make a "copy" of this branch (`git checkout -b newbranch` when you are starting on this one), and then do a rebase on master. If anything goes badly, you can always try again by going back to this branch.

ScottPJones commented:
@timholy Thanks! I'll try to get that fixed up... again, pardon my GitHub newbieness!

timholy commented:
You can also push branches with different names, e.g.,
```sh
git push remotename +localbranchname:remotebranchname
```
The `+` makes it a force-push, which you'll need after rebasing.

ScottPJones committed:
#11081 Add documentation about noteworthy differences between C/C++ and Julia

ScottPJones commented:
Ok folks, please review this!  I greatly welcome any and all constructive criticism!

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+- In Julia, indexing of arrays, strings, etc. is 1-based not 0-based.
+- Julia arrays are assigned by reference. After ``A=B``, changing elements of
+  ``B`` will modify ``A`` as well.
+- Julia values are passed and assigned by reference. If a function modifies an
```
I think values might be too vague. Its more like immutables are passed by value, and mutables are passed by reference, I believe.

ScottPJones replied:
Hmm... I just copied that section, from one of the previous sections (for MATLAB or R), because it was applicable to C/C++, however, I didn't think immutables were passed by value... although since they are immutable, it doesn't really matter, since you can't modify them in the called function anyway.

StefanKarpinski replied:
Everything is passed by sharing semantically, it's just that with immutables this is indistinguishable from passing by value. I'd prefer not to promote the mistaken idea that immutable and mutable values are treated differently.

ScottPJones replied:
Yes, I understood that, and that's why I left it as it was...

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  Unlike either decimal integer or hexadecimal literals, floating point literals are always
+  64-bits, Float64, and don't get promoted to the BigFloat type.  This is closer in
+  behavior to C/C++.
+- String literals can be delimited with either " or """
```
Might be worth mentioning that `"""` is there for multiline strings

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+- String literals can be delimited with either " or """
+  They can also have string interpolation, indicated by $variablename or $(exprssion),
+  which evaluates the variable name or the expression in the context of the function.
+- `//` indicates a Ration number, and not a single-line comment (which is # in Julia)
```
Rational

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  You cannot write ``if (1)`` in Julia, because ``if`` statements accept only
+  booleans. Instead, you can write ``if true``, ``if Bool(1)``, or ``if 1==1``.
+- In Julia, ``return`` does not require parentheses.
+- Julia uses ``end`` to denote the end of conditional blocks, like ``if``,
```
maybe just blocks, not just conditional blocks

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+- Julia uses ``end`` to denote the end of conditional blocks, like ``if``,
+  loop blocks, like ``while``/``for``, and functions. In lieu of the one-line
+  ``if ( cond ) statement``, Julia allows statements of the form
+  ``if cond; statement; end``, ``cond && statement`` and
```
semicolons aren't actually needed `if true println("true") end` although its not very readable. Might be worthing mentioning that the `&&` thing isn't special syntax, just a consequence of the boolean operator precedence.

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  ``if ( cond ) statement``, Julia allows statements of the form
+  ``if cond; statement; end``, ``cond && statement`` and
+  ``!cond || statement``. Assignment statements in the latter two syntaxes must
+  be explicitly wrapped in parentheses, e.g. ``cond && (x = value)``.
```
because of operator precedence

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  continues. One way to force an expression to continue is to wrap it in
+  parentheses.
+- By convention, functions that modify their arguments have a ! at the end of the name,
+  for example push!.
```
\`push!\`

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  They can also have string interpolation, indicated by $variablename or $(exprssion),
+  which evaluates the variable name or the expression in the context of the function.
+- `//` indicates a Ration number, and not a single-line comment (which is # in Julia)
+- `#=` indicates the start of a multiline comment, and =# ends it.
```
\`=#\`

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  keyword.  Multiple values can be returned from functions and assigned as tuples, e.g.
+  ``(a, b) = myfunction()`` or ``a, b = myfunction()``, instead of having to pass pointers
+  to values as one would have to do in C/C++ (i.e. ``a = myfunction(&b)``.
+- Julia discourages the used of semicolons to end statements. The results of
```
Might be worth mentioning that `;` suppresses output at the REPL, not completely clear from this bullet why `;` exists in Julia if you don't already know

tkelman replied:
Not so much "discourages the use of" as "does not require" - the presence or absence of a semicolon as the last character in a line only makes a difference in the repl

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,72 @@ Noteworthy differences from Python
+  statements are not automatically printed (except at the interactive prompt),
+  and lines of code do not need to end with semicolons. :func:`println` or
+  :func:`@printf` can be used to print specific output.
+- In Julia, the operator :obj:`$` performs the bitwise XOR operation, i.e. :obj:`^`
```
Could also be worth mentioning that `pow` is `^`, but maybe not necessary.

jiahao replied:
> `//` indicates a Ration number

Ration -> Rational. Also, should be double backquoted. =# is not formatted as a literal on the next line.

ScottPJones committed:
Update noteworthy-differences.rst

ScottPJones commented:
@IainNZ Hopefully this addressed most all of your concerns... I didn't mention about ; at the REPL, I didn't even know about that... I was just trying to document "gotchas" that would bother a C/C++ programmer.  Thanks so much for taking the time to review it!  (I also found a typo with max in a previous section, not mine, that I also fixed)

nalimilan commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,75 @@ Noteworthy differences from Python
+  and lines of code do not need to end with semicolons. :func:`println` or
+  :func:`@printf` can be used to print specific output.
+- In Julia, the operator :obj:`$` performs the bitwise XOR operation, i.e. :obj:`^`
+  in C/C++.  Also, these the bitwise operators do not have the same precedence as C/++.
```
Typo in `these the`.

ScottPJones replied:
Thanks very much!

ScottPJones committed:
Update noteworthy-differences.rst

IainNZ commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,75 @@ Noteworthy differences from Python
+- Julia arrays are indexed with square brackets, and can have more than one
+  dimension ``A[i,j]``.
+- In Julia, indexing of arrays, strings, etc. is 1-based not 0-based.
+- Julia arrays are assigned by reference. After ``A=B``, changing elements of
```
I guess saying `arrays` is probably too limited for C/C++ section, because everything is assigned by reference. I think it was appropriate in the MATLAB section because everything in MATLAB is basically an array.

ScottPJones replied:
OK, yes, so how do you best think it should be said?  (I really do NOT consider myself a good writer... [of code yes, of clear explanations, no!])
Thanks!

tkelman commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,75 @@ Noteworthy differences from Python
+  behavior to C/C++.
+- String literals can be delimited with either `"`  or `"""`, `"""` delimited literals can contain `"` characters without
+  quoting it like `"\\\""`
+  They can also have string interpolation, indicated by $variablename or $(expression),
```
It's not clear here whether "They" refers to only `"""` delimited strings. I'd say something like "String literals can have values of other variables or expressions interpolated into them, indicated by..."

Should also put ``` ` ``` around `$variablename` and `$(expression)`

ScottPJones replied:
Thanks!  I greatly appreciate the help!

jiahao replied:
In ReStructured text format, single backticks like

```
`~0x000f`.
```

is a formatting error. You should have double backticks.

tkelman commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,75 @@ Noteworthy differences from Python
+- Julia's ``->`` creates an anonymous function, it does not access a member via a pointer.
+- Julia does not require parentheses when writing ``if`` statements or
+  ``for``/``while`` loops: use ``for i in [1, 2, 3]`` instead of
+  ``for (i in c(1, 2, 3))`` and ``if i == 1`` instead of ``if (i == 1)``.
```
this first example is R syntax, right? probably not appropriate for this section

ScottPJones replied:
Oops!  I'd copied the part from R that looked pertinent to C/C++, and over copied!  Thanks!

ScottPJones committed:
Update noteworthy-differences.rst

ScottPJones commented:
Is there anything that still needs to be changed, before this could be merged?
I'd like to get this in to the 0.4 documentation, so that others coming from C/C++ don't suffer the same "gotchas" that I did when starting with Julia.

tkelman commented:
It looks like there are still comments above that haven't been addressed yet. Do you have local modifications that aren't pushed to this branch yet?

ScottPJones commented:
@tkelman, I thought I had addressed everything... I was editing directly here on GitHub, not pushing from a local copy.  What have I missed?  Thanks!

tkelman commented:
https://github.com/JuliaLang/julia/pull/11099#discussion_r29548924

https://github.com/JuliaLang/julia/pull/11099#discussion_r29556368

https://github.com/JuliaLang/julia/pull/11099#discussion_r29555429

I think rst needs double-backticks on code, and good to format all Julia type names that way too

ScottPJones committed:
Add double backtics for types, add extra backtics

ScottPJones committed:
Add explanation of ``>>>``, better explanation of ``;``

ScottPJones commented:
@tkelman Anything else?  I hope that's caught all of the issues... thanks for the heads up about double backtic, I saw somethings had `:obj:` or `:func:` and single backtics... didn't know what was correct.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+- In Julia, indexing of arrays, strings, etc. is 1-based not 0-based.
+- Julia arrays are assigned by reference. After ``A=B``, changing elements of
+  ``B`` will modify ``A`` as well.
+- Julia values are passed and assigned by reference. If a function modifies an
```
Technically, this isn't pass by reference – it's pass by sharing. This illuminates the difference a bit:

http://javadude.com/articles/passbyvalue.htm

This is all quite confusing since the meaning of the terminology seems to have changed over time.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+- Julia values are passed and assigned by reference. If a function modifies an
+  array, the changes will be visible in the caller.
+- In Julia, whitespace is significant, unlike C/C++, so care must be taken when adding/removing
+  whitespace from a Julia program.
```
This seems unduly scary. There are not a lot of situations where whitespace is significant. I would tone down the "scare" aspect of this.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  ``0x0`` and ``0x00`` have type UInt8, ``0x000`` and ``0x0000`` have type ``UInt16``, then
+  literals with 5 to 8 hex digits have type ``UInt32``, 9 to 16 hex digits type ``UInt64``, and more than
+  16 hex digits end up a signed ``BigInt`` type.  This needs to be taken into account when defining
+  hexadecimal masks, for example ``~0xf`` is very different from ``~0x000f``.
```
This would be clearer if you included the result value, i.e. `~0xf == 0xf0` is very different from `~0x000f == 0xfff0`.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  hexadecimal masks, for example ``~0xf`` is very different from ``~0x000f``.
+  Unlike either decimal integer or hexadecimal literals, floating point literals are always
+  64-bits, ``Float64``, and don't get promoted to the ``BigFloat`` type.  This is closer in
+  behavior to C/C++.
```
This isn't quite right since we have Float32 literals as well, but the "normal" floating-point literals do produce Float64.

ScottPJones replied:
Good point, I'll fix that up later (have to fix up the documenation for 0.4 anyway, because Jake merged in my change to prevent unsigned -> signed promotion)

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  characters without quoting it like ``"\\\""``
+  String literals can have values of other variables or expressions interpolated into them,
+  indicated by ``$variablename`` or ``$(expression)``, which evaluates the variable name or the expression in the context of the function.
+- ``//`` indicates a ``Rational`` number, and not a single-line comment (which is # in Julia)
```
You probably want to mark the `#` character up as code here.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  :func:`@printf` can be used to print specific output.
+  In the REPL, ``;`` can be used to suppress output.
+  ``;`` also has a different meaning within ``[ ]``, something to watch out for.
+  ``;`` can be used to separate 'statements' on a line, but are not strictly necessary in many cases,
```
Not clear why 'statements' is in single quotes. Also, since Julia is expression-oriented, it would seem better to consistently use "expression" rather than using "statement" some of the time.

ScottPJones replied:
Ok, good... I put it in quotes because I knew it wasn't *really* a statement... maybe it should also point out as you said that Julia is expression-oriented, not statement oriented as C/C++.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+- Julia's ``->`` creates an anonymous function, it does not access a member via a pointer.
+- Julia does not require parentheses when writing ``if`` statements or
+  ``for``/``while`` loops: use ``for i in [1, 2, 3]`` instead of
+  ``for (i in c(1, 2, 3))`` and ``if i == 1`` instead of ``if (i == 1)``.
```
You've still got R syntax in here.

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  far is a complete expression, it is considered done; otherwise the input
+  continues. One way to force an expression to continue is to wrap it in
+  parentheses.
+- Julia has an incredibly powerful macro facility, and macros are always indicated by the ``@``
```
Let's drop "incredibly". I think it's fairly credible to people coming from many languages.

ScottPJones replied:
Somehow I missed your comments (it showed up as "commented on an outdated diff"), but, coming from C/C++, which this section is about, I think "incredibly" *is* actually accurate ;-)
Many good C/C++ programmers discourage the use of macros entirely... and may discount Julia's macros because of that prejudice. Maybe this section should point to the Julia metaprogramming documentation?

pao replied:
A link there might be good. The point about the word "macro" having a different meaning to the C/C++ crowd is well-taken. Perhaps moving the more concrete description up, something more like:

"Julia macros operate on parsed expressions, rather than the text of the program, which allows them to perform sophisticated transformations of Julia code. Macro names start with the `@` character, and have both a function-like syntax `@blah(arg,arg)` and a statement-like syntax `@blah arg arg`."

etc.

The "Note:" bit could use a bit of prosification as well.

I'm happy to go ahead and make those changes if this seems reasonable?

jakebolewski replied:
go for it

ScottPJones replied:
Yes, much better expressed than I could have done, please do that!

pao replied:
#11162

StefanKarpinski commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+  Note: the second form can continue onto multiple lines, if the last expression on a line goes over to another line...
+  This can be a bit tricky to determine sometimes just where the macro ends.
+- Julia now has an enumeration type, expressed using the macro ``@Enum(name, value1, value2, ...)``
+  For example: ``@Enum(Fruit, Banana=1, Apple, Pear)``
```
The macro is lowercase: `@enum`.

carnaval commented:
Should this mention the arguably biggest difference with c++ that there is no static dispatch ? Every function is "virtual" wrt to the argument tuple.

Keno commented on doc/manual/noteworthy-differences.rst:
```diff
@@ -251,3 +251,94 @@ Noteworthy differences from Python
+- Julia does not treat the numbers ``0`` and ``1`` as Booleans.
+  You cannot write ``if (1)`` in Julia, because ``if`` statements accept only
+  booleans. Instead, you can write ``if true``, ``if Bool(1)``, or ``if 1==1``.
+- In Julia, ``return`` does not require parentheses.
```
Neither does it in C/C++.

ScottPJones replied:
Ah, again, overcopying on my part for the sections that seemed relevant from other languages... I'll fix.

ScottPJones committed:
Update noteworthy-differences.rst

Remove incorrect line about return not needing parenthesis, copied from the R section by mistake, copied/modified advice on column major vs. row major.

carnaval commented:
In C++ if you write something like
```C++
class A { int f() { return 0; } }
class B : A { int f() { return 1; } }
((A*)(new B))->f()
```
you will get 0. To have dynamic dispatch on the runtime type of B you have to annotate `f` as `virtual`. In julia every method is "virtual" (of course it's more general than that since we dispatch on every argument type, not only `this`, using the most-specific-declaration rule). I'm not quite sure how to put this concisely.

ScottPJones committed:
Update noteworthy-differences.rst

Added section (inspired by @carnaval) about C++ static dispatch vs. dynamic dispatch.

ScottPJones commented:
@carnaval Does that seem OK?  Hopefully this will be the last go-round!

tkelman commented:
this still had multiple un-addressed editorial comments on it

ScottPJones commented:
Since this has been merged and closed, when will it actually show up in the on-line documentation for latest?  I was surprised this morning that it hasn't shown up there... I had wanted to point my colleagues at it, as they have been running into some of the same issues...  Thanks!

StefanKarpinski commented:
Are you looking at docs for latest or release-0.3?

ScottPJones commented:
I did say above, latest.   Looked this morning.

pao commented:
I noticed it hadn't rendered yet yesterday; assumed RTD was behind on hook processing then forgot to check later. Probably need to kick RTD to regenerate the docs (I'd do it but I don't have permission) and make sure the webhook is still working?

StefanKarpinski commented:
Added you as a maintainer, @pao. There seems to be a "build problem", whatever that means for RTD.

pao commented:
Thanks @StefanKarpinski; I dropped some debug info in the issue linked above, will investigate tonight if no one tackles it in the next few hours.

pao commented:
Thanks for letting us know @ScottPJones, all sorted now.

ScottPJones commented:
Thanks so much!  I'm just happy to see some of my contributions make it in... (of course, for this, I plagiarized shamelessly from the other sections, and had great editing help!)
